### ISSUE TITLE
Training Code 

### ISSUE THREAD

**Issue by pashanitw** [gpt-omni/mini-omni#11]
Are you planning to release the training and finetuning code?

**Comment by nanowell** [comment]
![image](https://github.com/user-attachments/assets/818fde65-2dab-49e7-ab9a-1a1d184f5fd5)

**Comment by yukiarimo** [comment]
That’s sad

**Comment by nanowell** [comment]
I've been working with my own code for training a transformer Audio-to-Audio models, even before this project came out. I'm going to try and adapt my code to work with their model design and the way they break down audio data. My goal is to train a model directly on audio, without relying on text transcripts. They claim it's difficult for smaller models to understand the long-term patterns in audio, but I've already managed to do it with a massive audiobook dataset. I believe I can make it work with this new architecture too.
If I am successful I will release the code for training this architecture.

**Comment by nanowell** [comment]
Native Audio-to-Audio model with no tricks (no TTS, no Whisper or anything else) just a pure transformer based architecture (MH Casual Attention, basic MLP with gelu activation), Snac as tokenizer, I pretokenize dataset for faster training.

In my experience you will need aprox ~ hundreds of hours of consistent and high quality audio dataset ~100 millions of codec logits (in my case 32000 hz model). You will have to preprocess audio into single mono channel because snac does not support stereo, also checkout [WavTokenizer ](https://huggingface.co/novateur/WavTokenizer )which is OSS SOTA for audio tokenization (currently only speech model) that was specifically designed for gpt-4o type of models.

**Comment by mini-omni** [comment]
Hi, guys. We cannot guarantee that we can release the training code, so we don't want to give anyone false hope. However, with the inference code, the training is all about the data preparing and compute loss as the LLM does, if there are any implementation issues, we are glad to help.

**Comment by mini-omni** [comment]
> I've been working with my own code for training a transformer Audio-to-Audio models, even before this project came out. I'm going to try and adapt my code to work with their model design and the way they break down audio data. My goal is to train a model directly on audio, without relying on text transcripts. They claim it's difficult for smaller models to understand the long-term patterns in audio, but I've already managed to do it with a massive audiobook dataset. I believe I can make it work with this new architecture too. If I am successful I will release the code for training this architecture.

hi @nanowell , indeed, we found that if we predicted the audio tokens directly, the output was much worse than the parallel decoding. I think SpeechGPT and Spectron did speech-to-speech with QuestionAudio-QuestionText-AnswerText-AnswerAudio way, they may also found out that generating audio condtion on text was easier, and It makes sence intuitively. Of course, we will be excited to see you can make it work. Also ,we are glad to discuss more about it.

**Comment by kobenaxie** [comment]
> Hi, guys. We cannot guarantee that we can release the training code, so we don't want to give anyone false hope. However, with the inference code, the training is all about the data preparing and compute loss as the LLM does, if there are any implementation issues, we are glad to help.

hi @mini-omni , 有俩关于训练的问题请教一下：
1、在第一阶段的多模态对齐的过程中，论文里提到是用ASR和TTS数据训练模型的ASR和TTS能力，换句话说模型是同时支持ASR和TTS的，可是图3中第一阶段为什么只有语音的输入？没有文本的输入如何做TTS呢？
2、训练阶段的`task`是不是只有`A1|A2`,`A1|T2`,`T1|T2`,`T1|T2`四种情况？

**Comment by mini-omni** [comment]
> > Hi, guys. We cannot guarantee that we can release the training code, so we don't want to give anyone false hope. However, with the inference code, the training is all about the data preparing and compute loss as the LLM does, if there are any implementation issues, we are glad to help.
> 
> hi @mini-omni , 有俩关于训练的问题请教一下： 1、在第一阶段的多模态对齐的过程中，论文里提到是用ASR和TTS数据训练模型的ASR和TTS能力，换句话说模型是同时支持ASR和TTS的，可是图3中第一阶段为什么只有语音的输入？没有文本的输入如何做TTS呢？ 2、训练阶段的`task`是不是只有`A1|A2`,`A1|T2`,`T1|T2`,`T1|T2`四种情况？

hi, 
1、你观察很对，这里图示有些问题。asr, tts同时做adapter训练时，需要同时支持音频和文本输入。
2、ASR会有 A1T1，TTS则会有T1A1

**Comment by mini-omni** [comment]
I'll close it for now, please feel free to re-open.
